server/logs/lib.lua

```diff
@@ -9,33 +9,61 @@ function Logs.findContext()
 	return info
 end
 
-local function isValidLogLevel(level)
-	if level == "info" or "warn" or "error" then
-		return true
-	end
-
-	return false
-end
-
 local function sendHttpRequest(data)
 	PerformHttpRequest(apiUrl, function() end, "POST", json.encode(data), {
 		["Authorization"] = apiKey,
 		["Content-Type"] = "application/json",
 	})
 end
 
-function Logs:LogMessage(level, message, metadata)
-	local isValid = isValidLogLevel(level)
-	if not isValid then
-		print("Log level is not valid. Use 'info', 'warn' or 'error'.")
+local function getIdentifiers(playerSource)
+	local identifiers = GetPlayerIdentifiers(playerSource)
+	local data = {}
+	for i = 1, #identifiers do
+		--split the identifier at the colon
+		local identifier = identifiers[i]
+		local _, _, idType, id = string.find(identifier, "([^:]+):(.+)")
+		data[idType] = id
+	end
+
+	return data
+end
+
+---@alias logType string | '"info"' | '"warn"' | '"error"' | 'any' | string
+
+---@class LogOptions
+---@field metadata table @The metadata to send with the log. Can be an array or kv.
+---@field playerSource number @The player source to send with the log.
+
+---@param logType logType @The log level.
+---@param message string @The log message.
+---@param options LogOptions @The log options.
+function Logs:LogMessage(logType, message, options)
+	if not logType == type("string") or not message == type("string") then
+		print("Log type or message is missing.")
 		return
 	end
 
-	--self.findContext()
+	local player
+	if options?.playerSource then
+		if not type(options.playerSource) == type("number") then
```

Comment:
Few things here,

1. Going off of the metadata comment I left above, maybe instead of creating a "player", if the metadata contains a playerSource then you can just fetch the identifiers and attach them to the metadata object itself like in _identifiers or something similar? What are your thoughts on this alternative approach vs the initial approach? @itschip I'd like your thoughts as well.

2. I'm personally not a fan of using the power patches provided in the fivem Lua runtime, it basically forces all contributors to change their editors in someway, so I tend to stay away from things like optional chaining in Lua.

3. Like above I believe the type function is being used incorrectly here

4. It could also be a nicety to not overwrite the _identifiers property if the user passed it in themselves

Reply:
As commented above, I agree with point 1. This will keep metadata filtering universal and less work on the backend and database.